README.md
http-route-proxy
==========

Convenient HTTP/HTTPS route proxy for cross-domain request, request forward.Support using as express connect middleware.

### Installing

```bash
npm install http-route-proxy
```

### Testing

```bash
npm test
```

### Using http-route-proxy

#### Sample:

```javascript
var proxyServer = require('http-route-proxy');

/**
 *   proxy configs
 */
proxyServer.proxy([
    // common config
    {
        // origin host + port
        from: 'localhost:9000',
        // forward host + port
        to: '192.168.1.1:8088',
        // match forward action rule
        // `"/"` means forward match all actions, 
        // `!/public` means match local static forward match `/public.*`
        route: ['/', '!/public']
    }
]);
```

#### Set Headers
```javascript
proxyServer.proxy([
    {
        from: 'localhost:9000',
        to: 'www.google.com',
        https: true,
        headers: {
            req: {origin: 'www.google.com', referer: 'www.google.com'},
            res: {'access-control-allow-origin': 'https://www.google.com', 'access-control-allow-credentials': true}
        }
    }
]);
```

#### Using HTTPS
```javascript
proxyServer.proxy([
    {
        from: 'localhost:9000',
        to: 'www.google.com',
        https: true
    }
]);
// or
proxyServer.proxy([
    {
        from: 'localhost:9000',
        to: 'https://www.google.com'
    }
]);
```

#### Forward By Request Path
Forward to various host on matching the path action

```javascript
proxyServer.proxy([
    {
        from: 'localhost:9000',
        to: 'www.google.com',
        route: [
            {
                action: '/switer',
                forward: 'https://github.com',
                headers: {
                    req: {origin: 'https://github.com'}
                }
            },
            {
                action: '/imper',
                forward: 'https://switer.github.io',
                headers: {
                    req: {origin: 'https://github.io'}
                }
            },
            '!/public'
        ]
    }
]);
```

#### Using as express connect middleware
__express app.js config:__
Push it on the first of connect

```javascript
app.use(proxyServer.connect({
  to: 'www.google.com',
  https: true,
  route: ['/']
}));
app.use(express.favicon());
app.use(express.logger('dev'));
app.use(express.bodyParser());
app.use(express.methodOverride());
app.use(app.router);
app.use(express.static(path.join(__dirname, 'public')));
```

### Change Log

#### Version 0.2.0 - 2015/08/26
- [Bug]: Fixed bug when using as express middleware sometimes base-url will be stripped off. 

#### Version 0.1.1 - 2013/12/30

- [Feature]: Support using as express connect middleware

#### Version 0.1.0 - 2013/12/28

- [Feature]: Forward to various host on matching the path action.

#### Version 0.0.6-1 - 2013/12/26

- [Feature]: route field isn't necessary and it has defualt value of `['/']`

#### Version 0.0.6 - 2013/12/26

- [Feature]: Support set request and response headers

#### Version 0.0.5-1 - 2013/12/26

- [Fixed bug]: route proxy http to https website which certificate has not been unauthorized will be response error

#### Version 0.0.5 - 2013/12/25

- [Feature]: Support https route proxy

#### Version 0.0.4 - 2013/12/18

- [Feature]: Support cross-domain request


### License

The MIT License (MIT)

Copyright (c) 2013 `guankaishe`

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



CHANGELOG.md
## ChangeLog for: http-route-proxy

#### Version 0.2.0 - 2015/08/26
- [Bug]: Fixed bug when using as express middleware sometimes base-url will be stripped off. 

#### Version 0.1.1 - 2013/12/30

- [Feature]: Support using as express connect middleware

#### Version 0.1.0 - 2013/12/28

- [Feature]: Forward to various host on matching the path action.

#### Version 0.0.6-2 - 2013/12/26

- [Feature]: Support write protocal in hostname(https://xxxx.com) instead of using protocal(https: true) option.

#### Version 0.0.6-1 - 2013/12/26

- [Feature]: route field isn't necessary and it has defualt value of `['/']`

#### Version 0.0.6 - 2013/12/26

- [Feature]: Support set request and response headers

#### Version 0.0.5-1 - 2013/12/26

- [Fixed bug]: route proxy http to https website which certificate has not been unauthorized will be response error

#### Version 0.0.5 - 2013/12/25

- [Feature]: Support https route proxy

#### Version 0.0.4 - 2013/12/18

- [Feature]: Support cross-domain request


test/test.js
var proxyServer = require('../server');

/**
 *   proxy configs
 */
proxyServer.proxy([
    // {
    //     from: 'localhost:9004',
    //     to: 'https://www.google.com'
    // },
    {
        from: 'localhost:9004',
        to: 'https://kyfw.12306.cn'
    },
    {
        from: 'localhost:9003',
        route: [{
                action: '/',
                forward: 'https://www.google.com'
            },
            '!/public', // forward to local static files
            {
                action: '/otn',
                forward: 'localhost:9004',
                headers: {
                    req: {
                        origin: 'https://kyfw.12306.cn/',
                        referer: 'https://kyfw.12306.cn/'
                    }
                }
            }, {
                action: '/otn/login/init',
                forward: 'kyfw.12306.cn',
                headers: {
                    req: {
                        Origin: 'kyfw.12306.cn'
                    },
                    res: {
                        Doo: 'guankaishe'
                    }
                }
            }
        ]
    },
    // common config
    {
        // origin host + port
        from: 'localhost:9002',
        // forward host + port
        to: 'www.google.com',
        // protocal option
        https: true,
        // reset headers
        headers: {
            req: {
                origin: 'https://www.google.com',
                referer: 'https://www.google.com'
            },
            res: {
                Doo: 'www.google.com'
            }
        }
    },
    // host with 80 port
    {
        from: 'localhost:9001',
        to: 'github.com',
        https: true,
        headers: {
            req: {
                origin: 'https://github.com',
                referer: 'https://github.com'
            }
        }
    }
]);


test/https.js
var httpProxy = require('http-proxy');

var options = {
    changeOrigin: true,
    target: {
        https: true,
        rejectUnauthorized: false
    }
}

// process.env['NODE_TLS_REJECT_UNAUTHORIZED'] = '0';
httpProxy.createServer(options, function (req, res, proxy) {
    proxy.proxyRequest(req, res, {
        host: 'www.12306.cn',
        port: 443
    });
}).listen(9001);


test/index.js
var proxyServer = require('../index');

/**
 *   proxy configs
 */
proxyServer.proxy([
    // {
    //     from: 'localhost:9004',
    //     to: 'https://www.google.com'
    // },
    {
        from: 'localhost:9004',
        to: 'https://kyfw.12306.cn'
    },
    {
        from: 'localhost:9003',
        route: [{
                action: '/',
                forward: 'https://www.google.com'
            },
            '!/public', // forward to local static files
            {
                action: '/otn',
                forward: 'localhost:9004',
                headers: {
                    req: {
                        origin: 'https://kyfw.12306.cn/',
                        referer: 'https://kyfw.12306.cn/'
                    }
                }
            }, {
                action: '/otn/login/init',
                forward: 'kyfw.12306.cn',
                headers: {
                    req: {
                        Origin: 'kyfw.12306.cn'
                    },
                    res: {
                        Doo: 'guankaishe'
                    }
                }
            }
        ]
    },
    // common config
    {
        // origin host + port
        from: 'localhost:9002',
        // forward host + port
        to: 'www.google.com',
        // protocal option
        https: true,
        // reset headers
        headers: {
            req: {
                origin: 'https://www.google.com',
                referer: 'https://www.google.com'
            },
            res: {
                Doo: 'www.google.com'
            }
        }
    },
    // host with 80 port
    {
        from: 'localhost:9001',
        to: 'github.com',
        https: true,
        headers: {
            req: {
                origin: 'https://github.com',
                referer: 'https://github.com'
            }
        }
    }
]);


tools/static.js
var http = require('http'),
    send = require('send'),
    url = require('url'),
    colors = require('colors'),
    path = require('path');

var staticServer = {

    // server id with increasing
    serverId: 0,
    // serverr config with serverId as key
    server: {},

    init: function (hostname, port, options) {
        hostname = hostname || 'localhost';
        port = port || 9000;
        options = options || {};

        var directory = options.directory || '.',
            serverId = this.serverId ++;

        this.server[serverId] = {
            hostname: hostname,
            port: port,
            directory: directory
        };

        var _this = this;
        http.createServer(function(req, res) {
            // resolve root directory to local path 
            directory = path.resolve(process.cwd(), directory);
            // response file sending
            _this.sendfile(req, res, directory);

        }).listen(port, hostname);
        // marked server instance id
        return serverId;
    },
    /**
     *   send static file or directory which is requested
     */
    sendfile: function (req, res, directory) {
        // error handler
        function error(err) {
            res.statusCode = err.status || 500;
            err.message && console.log(err.message.red);
            res.end(err.message);
        }
        // directory redirect handler
        function redirect() {
            res.statusCode = 301;
            res.setHeader('Location', req.url + '/');
            res.end('Redirecting to ' + req.url + '/');
        }
        // send
        return send(req, url.parse(req.url).pathname)
            .root(directory)
            .on('error', error)
            .on('directory', redirect)
            .pipe(res);
    }
}

module.exports = staticServer;


index.js
/**
 *  http-route-proxy
 *  http://github.com/switer/http-route-proxy
 *
 *  Copyright (c) 2013 "switer" guankaishe
 *  Licensed under the MIT license.
 *  https://github.com/switer/http-route-proxy/blob/master/LICENSE
 */

var httpProxy = require('http-proxy'), // base on nodejitsu proxy server
    staticServer = require('./lib/static'), // static server
    util = require('./lib/util'),
    colors = require('colors'), // use to pretty console log
    path = require('path'),
    _ = require('underscore'); // use each method.....

var server = {

    /**
     *  request proxy route match status code
     */
    status: {

        STATIC: 0,
        FORWARD: 1,
        UNMATCHED: 2
    },
    /**
     *   id for each proxy server
     */
    serverId: 1,
    /**
     *   mapping root
     */
    rules: {},
    /**
     *   save proxy host
     */
    proxies: {
        /**
         *  from: Object {hostname, name}
         *  rules: Object {static, forward, rewrite}
         */
    },
    /**
     *   defualt port when config port is not pass in
     */
    defaultPort: 80,
    /**
     *  defualt port for https
     */
    defualtHttpsPort: 443,
    /**
     *   module regexps
     */
    regexps: {
        // validate host string
        HOST: /^[0-9a-zA-Z\.]+\:?[0-9]*$/,
        ROUTE_FORWARD: /^([A-Z]+:)?\/.*$/,
        // static request must be GET method
        ROUTE_STATIC: /^!([A-Z]+:)?\/.*$/
        // ROUTE_REWRITE: /^\^.*/
    },
    /**
     *  for creating request route matching rule
     */
    str: {
        ROUTE_METHOD_PREFIX: '([A-Z]+:)?'
    },
    /**
     *  Proxy enter api
     */
    proxy: function (hosts) {

        // handle each proxy config
        _.each(hosts, function (hostConfig) {

            var hostObj = {
                from: this.parseHost(hostConfig.from),
                routes: this.parseRouteRule(hostConfig)
            }

            var serverId = this.create(hostObj.from, hostObj.routes, hostConfig);
            this.saveHost(hostObj.from, hostObj.routes, serverId);

        }.bind(this));
    },
    /**
     *  support as express connect middleware
     */
    connect: function (config) {

        var proxy = new httpProxy.RoutingProxy(),
            connectConfig = {
                rules: this.parseRouteRule(config)
            },
            _this = this;

        return function (req, res, next) {

            _this.proxyMiddleware(req, res, proxy, connectConfig, function (proxyStatus) {
                // processing in express
                if (proxyStatus.status !== _this.status.FORWARD) {
                    next();
                }
                // forward request
                else {
                    console.log(proxyStatus.message);
                }
            });
        }
    },
    /**
     *   create a server for the one proxy config
     */
    create: function (from, routes, options) {

        var serverId = this.serverId, // for serverid hoisting
            // proxy server options
            proxyOptions = {
                // use client origin
                changeOrigin: true,
                target: {
                    // https: true,
                    // rejectUnauthorized: false
                }
            },
            _this = this; // for hoisting context of `this`
        // nessary for fixing node-http-proxy rejectUnauthorized option not work bug
        process.env['NODE_TLS_REJECT_UNAUTHORIZED'] = '0';
        // create proxy server
        var server = httpProxy.createServer(
            /**
             *  middleware
             *  Cross-Domain-Access
             */
            function (req, res, next) {
                // TODO
                next();
            }, 
            function (req, res, proxy) {

                _this.proxyMiddleware(req, res, proxy, _this.proxies[serverId], function (statusObj) {
                    console.log(statusObj.message);
                });
                
            /**
             *  must listen hostname, otherwise it will be fail when repeat listening 
             *  localhost in the some port
             */
            }).listen(from.port, from.hostname)
              .on('error', function (e) {
                // catch server listen error
                console.log('Create proxy error,'.red.grey + "cause by can't listen host from " + 
                    from.hostname.yellow.grey + ' : ' + from.port.toString().blue.grey);
              });
        
        console.log('Listen from ' + from.hostname.green.grey + ' : ' + from.port.toString().blue.grey);

        // Custom error
        server.proxy.on('proxyError', function (err, req, res) {
            res.writeHead(500, {
                'Content-Type': 'text/plain'
            });
            res.end('Something went wrong. And we are reporting a custom error message.\n' + err); 
        });

        return this.serverId ++;
    },
    /**
     *  Proxy middle for handling request and response
     */
    proxyMiddleware: function (req, res, proxy, config, next) {
        var from = this.parseHost(req.headers.host),
            method = req.method,
            // rewrite url to originUrl for proxy agent
            requestURL = req.url = req.originalUrl,
            url = method.toUpperCase() + ':' + req.originalUrl,
            forwardRouteObj = null;

        // get proxy config by proxy server id
        var proxyConfig = config;

        if (this.staticMatched(url, proxyConfig.rules.static)) {
            // match route is static file response
            var directory = path.resolve(process.cwd(), '.');

            // send static files without server
            staticServer.sendfile(req, res, directory);

            next({
                status: this.status.STATIC,
                message: method.blue.grey + '  ' + requestURL + ' from '.green.grey +  from.hostname + ':' + from.port.toString().blue.grey
            });
            
        } else if (forwardRouteObj = this.forwardMatched(url, proxyConfig.rules.forward)) {
            // set headers of config
            this.setHeaders(req, res, forwardRouteObj.options.headers);

            var forwardObj = forwardRouteObj.forward,
                // forward options
                proxyForwardOptions = {
                    changeOrigin: true,
                    host: forwardObj.hostname,
                    port: forwardObj.port
                };

            if (forwardRouteObj.options.https) {
                // set https forward options
                proxyForwardOptions = _.extend(proxyForwardOptions, {
                    target: {
                        https: true,
                        rejectUnauthorized: false
                    }
                });
            }

            // forward to remote server
            proxy.proxyRequest(req, res, proxyForwardOptions);

            next({
                status: this.status.FORWARD,
                message: 'forward '.yellow.grey + method.blue.grey + '  ' + requestURL + ' from '.green.grey +  from.hostname + ':' + 
                    from.port.toString().blue.grey +' to '.green.grey + forwardObj.protocal + '://' + forwardObj.hostname + ':' + 
                    forwardObj.port.toString().blue.grey + requestURL
            });
        }
        else {
            next({
                status: this.status.UNMATCHED,
                message: "Sorry proxy error, http-route-proxy can't match any forward rule, please check your proxy config".red
            });
        }
    },
    /**
     *  Set headers by proxy config
     */
    setHeaders: function (req, res, headers) {
        var reqHeaders = headers && headers.req ? headers.req: {},
            resHeaders = headers && headers.res ? headers.res: {},
            origin = req.headers.origin || req.headers.host;

        // if exist Origin, use "Cross-Domain-Access"
        if (origin) {
            // currently, we only support receiving http request
            res.setHeader('access-control-allow-origin', 'http://' + origin);
            res.setHeader('access-control-allow-credentials', true);
        }
        //rewrite header on config
        if (headers) {
            _.each(reqHeaders, function (value, key) {
                req.headers[key] = value;
            });
            _.each(resHeaders, function (value, key) {
                res.setHeader(key, value);
            });
        }
        // avoid node-http-proxy to rewrite headers
        var setHeader = res.setHeader;
        res.setHeader = function (key, value) {
            if (!resHeaders[key]) setHeader.call(res, key, value);
        }

    },
    /**
     *   host is a string, should be parse to hostname + port object format
     */
    parseHost: function (host, options) {
        options = options || {};

        var hostChunks = host.match(/^([a-z]+)\:\/\/(.*)$/),
            protocal = hostChunks? hostChunks[1]: '',
            hostSection = hostChunks? hostChunks[2]: host,
            hostname = hostSection.split(':')[0],
            port = hostSection.split(':')[1],
            protocalObj = this.protocal(protocal),
            protocalOptions = protocalObj.unknow ? options: protocalObj;

        return {
            protocal: this.options2protocal(protocalOptions),
            protocalOption: protocalOptions,
            hostname: hostname,
            port: port ? parseInt(port) : (protocalOptions.https ? this.defualtHttpsPort : this.defaultPort)
        };
    },
    /**
     *   save each host config
     */
    saveHost: function (fromObj, routeRules, serverId) {

        // save proxy config
        this.proxies[serverId] = {

            from: fromObj,
            rules: routeRules

        };

    },
    /**
     *  Check route if has default action , if has not, put it on first
     */
    checkDefaultAction: function (routes) {

        var isContainDefaultAction = false;

        _.each(routes, function (item) {
            if (item == '/' || item.action == '/') {
                isContainDefaultAction = true;
                return true;
            }
        });
        // put default action on the first tuples
        if (!isContainDefaultAction) routes.unshift('/'); 
        return routes;
    },
    /**
     *   validate host string
     */
    validHost: function (host) {
        return this.regexps.HOST.exec(host) ? true : false;
    },
    /**
     *  
     *  @return Object
     */
    protocal: function (protocal) {
        var protocalObj = {
            http: false,
            https: false,
            websocket: false,
            unknow: false
        };
        if (protocal === 'http') {
            protocalObj.http = true;
        } else if (protocal === 'https') {
            protocalObj.https = true;
        } else if (protocal === 'ws') {
            protocalObj.websocket = true;
        } else {
            protocalObj.unknow = true;
        }
        return protocalObj;
    },
    /**
     *  protocal option to protocal name
     */
    options2protocal: function (options) {

        if (options.https) {
            return 'https';
        } else if (options.websocket) {
            return 'ws';
        } else if (options.http) {
            return 'http';
        } else {
            return 'http';
        }
    },
    /**
     *   parse each route rule to url matched rule
     */
    parseRouteRule: function (options) {
        options = util.copy(options);

        var routes = options.route || [];
        // if route config is empty, give it default
        routes = this.checkDefaultAction(routes);

        var _this = this,
            forwards = [], // forward route rules
            // rewrites = [], // currently it's useless
            statices = [], // static route rules
            forwardObj = null;

        _.each(routes, function (route) {

            // route detail config 
            if (_.isObject(route)) {

                var routeOptions = route,
                    forwardObj = _this.parseHost(routeOptions.forward);

                routeOptions = _.extend(routeOptions, forwardObj.protocalOption);

                forwards.push({
                    rule: _this.forwardRouteRule2Regexp(routeOptions.action),
                    forward: forwardObj,
                    options: routeOptions
                });

            }
            // forward route match, route rule in shorthand
            else if (_this.regexps.ROUTE_FORWARD.exec(route)) {
                // parse forward host
                var forwardObj = _this.parseHost(options.to, options);
                // set foward options
                options = _.extend(options, forwardObj.protocalOption);
                // save config
                forwards.push({
                    rule: _this.forwardRouteRule2Regexp(route),
                    forward: forwardObj,
                    options: options
                });

            }
            // static route match, route rule in shorthand
            else if (_this.regexps.ROUTE_STATIC.exec(route)) {
                statices.push(_this.staticRouteRule2Regexp(route));
            }
        });

        return {
            forward: forwards,
            // rewrite: rewrites,
            static: statices
        };
    },
    /**
     *   url route rule to regexp
     */
    string2Regexp: function (rule) {
        rule = rule.replace('/','\\/');
        rule = '^' + rule + '.*$';
        return new RegExp(rule);
    },
    /**
     *  find the rule whether is static route rule and return regexpess rule.
     */
    staticRouteRule2Regexp: function (rule) {
        var matches = this.regexps.ROUTE_STATIC.exec(rule);

        rule = rule.replace(/^!/, '');

        if (!matches[1]) {
            rule = this.str.ROUTE_METHOD_PREFIX + rule;
        }
        return this.string2Regexp(rule);
    },
    /**
     *  find the rule whether is forward route rule and return regexpess rule.
     */
    forwardRouteRule2Regexp: function (rule) {
        var matches = this.regexps.ROUTE_FORWARD.exec(rule);

        if (!matches[1]) {
            rule = this.str.ROUTE_METHOD_PREFIX + '(@)'.replace('@', rule);
        }
        return this.string2Regexp(rule);
    },
    /**
     *  Match url by forward route rule
     *  @return {Object} routeRule
     */
    forwardMatched: function (url, forwardRules) {
        var matchedRoute = null;

        _.each(forwardRules, function (ruleObj) {
            if (ruleObj.rule.exec(url)) {
                matchedRoute = ruleObj;
                return true;
            }
        });

        return matchedRoute;
    },
    /**
     *  Match url by static route rule
     *  @return <Boolean> isMatched
     */
    staticMatched: function (url, staticRules) {
        var isMatched = false;

        _.each(staticRules, function (rule) {
            if (rule.exec(url)) {
                isMatched = true;
                return true;
            }
        });

        return isMatched;
    }
}

module.exports = server;